File: clrsPython/Lecture_1/Binary_Search.py
def Iterative_Binary_Search(A,v,n):
    ## A[1...n] is the sorted array
    ## v is the query number
    ## n is the length of the array
    l = 0
    r = n -1
    while l <= r:
        m = int(((l+r)/2))
        if A[m] == v:
            return m
        elif A[m] < v:
            l = m + 1
        else:
            r = m -1
    return None

def Recursive_Binary_search(A,v,l,r):
    ## A[1...n] is the sorted array
    ## v is the query number
    ## l,r are left and right bounds, equal 0 and n-1 in initialization
    if l > r:
        return None
    m = int(((l + r) / 2))
    if A[m] == v:
        return m
    elif A[m] < v:
        return Recursive_Binary_search(A,v,m+1,r)
    else:
        return Recursive_Binary_search(A,v,l,m-1)

def Search_2_elems_equal_x(S,x,n):
    for i in S:
        pos = Iterative_Binary_Search(S,x-i,n)
        if pos != None:
            return i, pos
    return None


A = [1,3,4,4,5,6,6,7,9,9.7,10,11,11.3,14]
v = 9
print(Iterative_Binary_Search(A,v,len(A)))
print(Recursive_Binary_search(A,v,0,len(A)-1))

S = [1,2,3,4,5,6,7,8,9]
x = 6
print(Search_2_elems_equal_x(S,x,len(S)))

File: clrsCpp/lecture_1/Binary_Search.cpp
#include <iostream>

int Iterative_Binary_Search(int* array, int v, int n)
{
	// A[1...n] is the sorted array
	// v is the query number
	// n is the length of the array
	int l = 0;
	int r = n - 1;
	while (l <= r)
	{
		int m = floor((l + r) / 2);
		if (array[m] == v)
		{
			return m;
		}
		else if (array[m] < v) // l < m < v < r
		{
			l = m + 1;
		}
		else // l < v < m < r
		{
			r = m - 1;
		}
	}
	return -1;
}


int Recursive_Binary_Search(int* array, int v, int l, int r)
{
	if (l > r)
	{
		return -1;
	}
	else
	{
		int m = floor((l + r) / 2);
		if (array[m] == v)
		{
			return m;
		}
		else if (array[m] < v)
		{
			return Recursive_Binary_Search(array, v, m + 1, r);
		}
		else
		{
			return Recursive_Binary_Search(array, v, l, m - 1);
		}
	}
}


int* Search_two_elements_equal_x(int* S, int x, int n)
{
	for (int i = 0; i < n; i++)
	{
		int pos = Iterative_Binary_Search(S, x, n);
		if (pos != -1)
		{
			int array[2] = { i,pos };
			return array;
		}
	}
	int array[2] = { -1,-1 };
	return array;
}